### PR TITLE
Replace vroom with data.table::fread

### LINKED ISSUE #198: Explore even more efficient CSV reading
The CSV reading is now quite fast after we replaced `utils::read_csv()` with vroom. We could however use vroom even more efficiently. Currently the altrep argument is set to FALSE [here](https://github.com/stan-dev/cmdstanr/blob/master/R/read_csv.R#L165). 

The default option is TRUE, which indexes the CSV once reads the data lazily as you need it. We currently index and read the data immediately and repeat the process every time we read the same CSV. 

I tried using this in the original vroom PR, but the issue was that I was not able to delete the CSVs in the same session. Which means that the `fit$save_output_files()` failed.

There is an issue on the vroom package: https://github.com/r-lib/vroom/issues/177 
It was closed, though I dont think this was resolved. There might be other workarounds worth exploring.

I might try playing with this again at a later point, but if someone wants to play with this, you are more then welcome to.

### LINKED ISSUE #262: Replace vroom with fread
**Describe the issue**

~We should move `vroom` from `Imports` to `Suggests` in the `DESCRIPTION` file. This was suggested by [Michael Peck on Discourse](https://discourse.mc-stan.org/t/lightweight-interfaces-keeping-it-light/16932/11?u=jonah). It has a lot of dependencies, so it's best to use it if present but not force people to install it.~

edit by Rok: 

We can replace vroom with data.table's fread function which makes making vroom optional redundant.

**CmdStanR version number**

0.1.0

### LINKED ISSUE #299: read_cmdstan_csv efficiency issue for large number of parameters
**Describe the bug**
CSV read speed drastically degrades when the number of params is large(>10^5) .

**To Reproduce**
@billgillespie and I ran into a model with 10^6 params and `read_cmdstan_csv` becomes stalled . I was able to reproduce it with the following dummy model
```stan
parameters {
  real k;
}

transformed parameters {
  real x[100000] = rep_array(k, 100000);
}

model {
  k ~ normal(0, 1);
}
```
The output file contains only one sample:
```bash
./dummy sample algorithm=fixed_param num_warmup=0 num_samples=1 output file=dummy.csv
```

`rstan::read_stan_csv` vs `cmdstanr::read_cmdstan_csv`:
```r
> system.time(fit <- cmdstanr::read_cmdstan_csv("dummy.csv"))
   user  system elapsed 
 55.873  26.429  67.056 
> system.time(fit <- rstan::read_stan_csv("dummy.csv"))
   user  system elapsed 
  5.449   0.341   5.847 
```

**Expected behavior**
More realistic reading speed in this scenario.

**Operating system**
macos mojave 
```r
> sessionInfo()
R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.6.2 tools_3.6.2    tcltk_3.6.2   
```

**CmdStanR version number**
```r
> library("cmdstanr")
This is cmdstanr version 0.1.3
- Online documentation and vignettes at mc-stan.org/cmdstanr
- Use set_cmdstan_path() to set the path to CmdStan
- Use install_cmdstan() to install CmdStan
```

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Summary

Replaces vroom with data.table

Closes #299 
closes #198 
closes #262 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Univ. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/data.R
Comment (jgabry) [inline R/data.R:166]: Two things: 

1. Why do we need to remove the names here? That's fine if necessary, just curious. 

2. Right now this isn't assigned to anything. I think you need 

```
chain_draws <- unname(chain_draws)
```

File: tests/testthat/test-fit-mcmc.R
Comment (jgabry) [inline tests/testthat/test-fit-mcmc.R:89]: Are all these lines commented out on purpose?

**Review by jgabry — COMMENTED** [review]
Thanks for this, looks good! I made a few tiny comments, but otherwise the main thing is that it's a bit difficult for me to parse (pun intended) the changes to the `read_csv_metadata()` function from the diff. It's a bit hard to follow. Are there particular lines that I should focus on checking? 

Also do you know why the windows unit tests are failing but not mac or linux?

File: R/data.R
Comment (rok-cesnovar) [reply R/data.R:166]: > Why do we need to remove the names here? That's fine if necessary, just curious.

Because it otherwise writes the iteration ids in the CSV.

>Right now this isn't assigned to anything. I think you need

Hm, this did help with that, will double-check.

File: tests/testthat/test-fit-mcmc.R
Comment (rok-cesnovar) [reply tests/testthat/test-fit-mcmc.R:89]: This test fails on macOS machine on the CI. I am currently unable to debug further. If you have a few minutes, mind running this test if it fails for you?

**Comment by rok-cesnovar** [comment]
> Are there particular lines that I should focus on checking?

Will explain a bit more once I get all tests passing. The main idea is that we read in all "metadata" lines and parse them then. Previously we read metadata lines line-by-line.

> Also do you know why the windows unit tests are failing but not mac or linux?

Its having problems finding grep. grep is a part of the Rtools installation so this is just a CI configuration thing I believe.

File: tests/testthat/test-fit-mcmc.R
Comment (jgabry) [reply tests/testthat/test-fit-mcmc.R:89]: Yeah I can check in a few min and let you know

Comment (jgabry) [reply tests/testthat/test-fit-mcmc.R:89]: This test passes on my mac

Comment (rok-cesnovar) [reply tests/testthat/test-fit-mcmc.R:89]: thanks! Thats good news and bad news at the same time :)

Comment (jgabry) [reply tests/testthat/test-fit-mcmc.R:89]: yeah good and bad. maybe we should let it run on CI again and I can see if I can debug it

**Comment by rok-cesnovar** [comment]
Status update: still issues with tests. I can use cmdstanr with fread on Windows locally (run a model and read in draws works fine) but cant get tests to pass both locally or in CI. Non-windows side of thing is ready, however.

**Comment by jgabry** [comment]
Thanks for the update. Is the problem on windows still related to failing to find "grep"?

**Comment by rok-cesnovar** [comment]
Seems so. Though it seems to find it in “normal use” but not when running tests. Should have time look into this a bit more this week and finish this up.

**Comment by jgabry** [comment]
That's strange. Sorry for the hassle, that sounds annoying, but thanks for working on this!

**Comment by mike-lawrence** [comment]
You don't have Cygwin installed on your local Windows machine, do you? [Here's someone](https://stackoverflow.com/questions/54661559/using-rtools-grep-pipe-combination-through-a-system-call) similarly having trouble with the rtools-installed grep, and I see that the [data.table maintainer seems to suggest ](https://stackoverflow.com/questions/36256706/fread-together-with-grepl/36270543) installing grep via Cygwin as the primary way to get it working.

**Review by rok-cesnovar — COMMENTED** [review]
Eureka! 

I think I solved the Windows CI issue and this is ready for review and merge finally. That took way too long.

Posting a few comments to make the review easier. Apart from replacing vroom the biggest change is the refactored metadata read.

**Comment by rok-cesnovar** [comment]
>  Here's someone similarly having trouble with the rtools-installed grep, and I see that the data.table maintainer seems to suggest installing grep via Cygwin as the primary way to get it working.

Thanks, @mike-lawrence for the suggestion. But even if this turned out to work, that would not be doable for cmdstanr as it would require users to also have Cygwin installed. I think I managed to find a workaround that works for Windows locally and in CI.

**Comment by rok-cesnovar** [comment]
I will run some benchmarks tomorrow to double check everything.

**Comment by jgabry** [comment]
Awesome! I'll try to review this soon.

**Review by jgabry — APPROVED** [review]
Curious to see what the benchmarks say but aside from that this is great and I think it's ready to merge! I'll approve it now and then assuming the benchmarks look good we can go ahead and merge. 

I think we also said we would tag version 0.2.0 after merging this, right? If that still sounds good to you then after this is merged we should update the DESCRIPTION file, the website, and upload the source package to stan-dev/r-packages. I'm happy to take care of that.

**Comment by rok-cesnovar** [comment]
Benchmark results of :
- this branch 
- rstan::read_stan_csv
- fread + as_draws_array (this is our floor that we can't reach because we need to handle metadata)

I did not compare with our current implementation as we know it's slower by a lot for large cases. And speedup is not the only reason we want to replace vroom with data.table.

All runs were done with 500 iterations and this model:
```stan
data {
  int N;
}
parameters {
  real k;
}
transformed parameters {
  real x[N] = rep_array(k, N);
}
model {
  k ~ normal(0, 1);
}
```
freadvsutilsnonlog

log scale:
freadvsutils

So this all looks good. One thing I am not exactly sure and can't explain is that in the `fread + as_draws_array` case, half of the time is spent in the as_draws_array.

```R
f <- data.table::fread(cmd= paste0("grep -v '^#' ", fit$output_files()[1]))
d <- posterior::as_draws_array(f)
```
So for the case of N=150000, its 5 seconds for fread and 4 seconds for as_draws_array. The latter seems a lot for that, but maybe I am missing something. Anyhow that is definitely something to figure out separately, not directly connected to this PR.

**Comment by rok-cesnovar** [comment]
For 0.2.0 I agree to do it after this is merged. I will also open another simple PR today to clean up some other issues, but none of them are critical.

**Comment by jgabry** [comment]
> So this all looks good. One thing I am not exactly sure and can't explain is that in the `fread + as_draws_array` case, half of the time is spent in the as_draws_array.

Ok yeah that looks good! I'm also surprised that half the time goes to creating the draws array. That does seem strange but I agree that's something to sort out separately so we can go ahead and merge this.

**Comment by jgabry** [comment]
Ok I'm going to merge this now and then do 0.2.0!

**Pull request merged**
